pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -171,6 +174,86 @@ func referencesDiffs(old []metatypes.OwnerReference, new []metatypes.OwnerRefere
 	return added, removed
 }
 
+func shouldFinalize(e event, accessor meta.Object) bool {
+	// The delta_fifo may combine the creation and update of the object into one
+	// event, so we need to check AddEvent as well.
+	if e.oldObj == nil {
+		if accessor.GetDeletionTimestamp() == nil {
+			return false
+		}
+	} else {
+		oldAccessor, err := meta.Accessor(e.oldObj)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
+			return false
+		}
+		// ignore the event if it's not updating DeletionTimestamp from non-nil to nil.
+		if accessor.GetDeletionTimestamp() == nil || oldAccessor.GetDeletionTimestamp() != nil {
+			return false
+		}
+	}
+	return sets.NewString(accessor.GetFinalizers()...).Has(OrphanFinalizerID)
+}
+
+func orphan(gc *GarbageCollector, owner *node) {
+	// TODO: This requires lock
+	dependents := owner.dependents
+	for dependent := range dependents {
+		deleteOwnerRefPatch := fmt.Sprintf(`{"metadata":{"ownerReferences":[{"$patch":"delete","uid":"%s"}]}}`, owner.identity.UID)
+		_, err := gc.patchObject(dependent.identity, []byte(deleteOwnerRefPatch))
+		if err != nil {
+			glog.V(6).Infof("orphaning %s failed with %v", dependent.identity, err)
+		}
+	}
+	// update the owner, remove "orphaningFinalizer" from its finalizers list
+	// TODO: maybe we should retry infinitely?
+	// TODO: Using Patch when strategicmerge supports deleting an entry from a
+	// slice of a base type.
+	maxRetries := 10
+	count := 0
+	for {
```

Comment:
Done. Returned an error message indicating the user to orphaning and finalizing manually.